aep/general/0122/aep.md

```diff
@@ -0,0 +1,318 @@
+
+Resource paths are formatted according to the [URI path schema][], but without
+the leading slash:
+
+    publishers/123/books/les-miserables
+    users/vhugo1802
+
+- Resource path components **should** usually alternate between collection
+  identifiers (example: `publishers`, `books`, `users`) and resource IDs
+  (example: `123`, `les-miserables`, `vhugo1802`).
+- Resource paths **must** use the `/` character to separate individual segments
+  of the resource path.
+  - Non-terminal segments of a resource path **must not** contain a `/`
+    character.
+  - The terminal segment of a resource path **should not** contain a `/`
+    character.
+- Resource paths **should** only use characters available in DNS names, as
+  defined by [RFC-1123](https://tools.ietf.org/html/rfc1123).
+  - Additionally, resource IDs **should not** use upper-case letters.
+  - If additional characters are necessary, resource paths **should not** use
+    characters that require URL-escaping, or characters outside of ASCII.
+  - If Unicode characters can not be avoided, resource paths **must** be stored
+    in Normalization Form C (see [AIP-210][]).
+- Eacg resource **must** expose a `path` field that contains its resource path.
+  - Resources **may** provide the resource ID, i.e. the last segment of the
+    path, as a separate field named `id`.
+  - Resources **may** expose a separate, system-generated unique ID field
+    (`uid`).
+  - Resources **must not** expose tuples, self-links, or other forms of
+    resource identification.
+  - All ID fields **must** be strings.
+
+**Note:** Resource paths as described here are used within the scope of a
+single API (or else in situations where the owning API is clear from the
+context), and are only required to be unique within that scope. For this
+reason, they are sometimes called _relative resource paths_ to distinguish them
+from _full resource paths_ (discussed below).
+
+[aip-210]: ./0210.md
+[uri path schema]: https://datatracker.ietf.org/doc/html/rfc3986#appendix-A
+
+### Collection identifiers
+
+The collection identifier segments in a resource path **must** be the plural
+form of the noun used for the resource. (For example, a collection of
+`Publisher` resources is called `publishers` in the resource path.)
+
+- Collection identifiers **must** be concise American English terms.
+- Collection identifiers **must** be in `kebab-case`.
+- Collection identifiers **must** begin with a lower-cased letter and contain
+  only lower-case ASCII letters, numbers. and hyphens (`/[a-z][a-z0-9-]*/`).
+- Collection identifiers **must** be plural.
+  - In situations where there is no plural word ("info"), or where the singular
+    and plural terms are the same ("moose"), the non-pluralized (singular) form
+    is correct. Collection segments **must not** "coin" words by adding "s" in
+    such cases (e.g. avoid "infos").
+
+#### Nested collections
+
+If a resource path contains multiple levels of a hierarchy, and a parent
```

Comment:
The resource has annotations that provide the path.

Here's a [real example](https://create.roblox.com/docs/en-us/reference/cloud/datastores-api/ordered-v1#GET-v1-universes-_universeId_-orderedDataStores-_orderedDataStore_-scopes-_scope_-entries-_entry_) from Roblox:  `/v1/universes/{universeId}/orderedDataStores/{orderedDataStore}/scopes/{scope}/entries/{entry}`.

If I had told the team that the AIPs required  `/v1/universes/{universeId}/orderedDataStores/{orderedDataStore}/orderedDataStoreScopes/{scope}/orderedDataStoreEntries/{entry}`, the understandable response would have been that the AIPs are causing more harm than good, making the APIs ugly and confusing, etc.

Reply:
Discussed live - needs group discussion.

Reply:
...and @toumorokoshi will try to come up with an example that benefits from removing this.

Reply:
Per @hudlow:  Maybe add a line about how organizations may *ban* this guidance, and then have lint rules both ways, so that organizations can disable one of those rules?

Reply:
@hudlow also suggests resource reference support union typing (or that we have reference interfaces that do the same thing in effect).  This is relevant to this topic in case we want to enforce a single path pattern for each resource so that each resource can be named according to its full resource hierarchy, meaning that a shared resource like `Log` (see google.aip.dev/4231) would have to be top-level, using some kind of generic (but ideally union-typed) resource reference for its parent.  (An alternative would be a oneof of standard resource references, but this also has drawbacks).

I'm not sure I agree with Dan's overall goal on that point, but it's worth discussing and I certainly think some of his points are valid.

Reply:
More generally, there was consensus from the attendees today that the abbreviated form is better (the alternative being "impossible to look at"), and interest in better understanding your objections @toumorokoshi.  Attendees included @Alfus, @hudlow, Andrei, Victor, and @makahmad.

Reply:
I've agreed to compromise on this in the meeting, so don't make this a blocker. There was general consensus in the sync that this change is desirable, and perhaps this should be a **must**.